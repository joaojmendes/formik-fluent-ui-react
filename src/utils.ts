import { FieldProps, getIn } from 'formik'

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>

export function getErrorMessage<T = any>({ field, form }: FieldProps<T>) {
  const error = getIn(form.errors, field.name)
  const touched = getIn(form.touched, field.name)

  return touched ? error : undefined
}

export function invokeAll(...callbacks: any[]) {
  return () => {
    for (const callback of callbacks) {
      if (callback && typeof callback === 'function') {
        callback()
      }
    }
  }
}

export function createFakeEvent({ name }: { name: string }) {
  return { target: { name } }
}
