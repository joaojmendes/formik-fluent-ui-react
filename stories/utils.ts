import { action } from '@storybook/addon-actions'
import { FormikActions } from 'formik'

export function handleSubmit<T>(
  values: T,
  { setSubmitting }: FormikActions<T>
) {
  action('submit:start')(values)
  setTimeout(() => {
    setSubmitting(false)
    action('submit:done')()
  }, 2000)
}
