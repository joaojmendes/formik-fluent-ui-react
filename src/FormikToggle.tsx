import { IToggleProps, Toggle } from '@fluentui/react';
import { FieldProps } from 'formik'
import * as React from 'react'
import { createFakeEvent, Omit } from './utils'

export function mapFieldToToggle<T = any>({
  form,
  field,
}: FieldProps<T>): Pick<IToggleProps, 'checked' | 'onChange'> {
  return {
    onChange: (_, checked) => {
      form.setFieldValue(field.name, checked)
      field.onBlur(createFakeEvent(field))
    },
    checked: field.value,
  }
}

export type FormikToggleProps<T = any> = Omit<
  IToggleProps,
  'checked' | 'onChange' | 'onBlur'
> &
  FieldProps<T>

export function FormikToggle<T = any>({
  field,
  form,
  ...props
}: FormikToggleProps<T>) {
  return <Toggle {...props} {...mapFieldToToggle({ field, form })} />
}
