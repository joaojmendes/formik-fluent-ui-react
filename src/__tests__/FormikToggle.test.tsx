import { Field, FieldProps, Form, Formik } from 'formik'
import { Toggle } from 'office-ui-fabric-react'
import * as React from 'react'
import renderer from 'react-test-renderer'
import { FormikToggle, mapFieldToToggle } from '../FormikToggle'
import { noop, serialize } from './utils'

class Values {
  public isChecked: boolean = false
}
function createFieldProps(value: boolean = false): FieldProps<Values> {
  return {
    field: {
      value,
      onChange: jest.fn(),
      onBlur: jest.fn(),
      name: 'isChecked',
    },
    form: { setFieldValue: jest.fn(), handleBlur: jest.fn(() => jest.fn()) },
  } as any
}

test('<FormikToggle /> renders correctly as a field component', () => {
  const component = renderer.create(
    <Formik initialValues={new Values()} onSubmit={noop}>
      <Form>
        <Field name="isChecked" label="IsChecked" component={FormikToggle} />
      </Form>
    </Formik>
  )

  expect(component.toJSON()).toMatchSnapshot()
})

test('<FormikToggle /> renders a Fabric <Toggle />', () => {
  const label = 'Date'
  const fieldProps = createFieldProps()

  const formikToggle = renderer.create(<FormikToggle {...fieldProps} />)
  const fabricToggle = renderer.create(
    <Toggle {...mapFieldToToggle(fieldProps)} label={label} />
  )
  expect(serialize(formikToggle)).toBe(serialize(fabricToggle))
})

test('mapFieldToToggle() maps FieldProps to IToggleProps', () => {
  const { field, form } = createFieldProps()
  const props = mapFieldToToggle({ form, field })

  expect(props.checked).toBe(field.value)

  props.onChange!(null as any, true)

  expect(form.setFieldValue).toHaveBeenCalledTimes(1)
  expect(form.setFieldValue).toHaveBeenCalledWith(field.name, true)

  // onChange also calls onBlur
  expect(field.onBlur).toHaveBeenCalled()
})
