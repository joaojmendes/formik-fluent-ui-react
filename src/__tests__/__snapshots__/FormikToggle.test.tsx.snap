// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormikToggle /> renders correctly as a field component 1`] = `
<form
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="ms-Toggle is-enabled root-47"
  >
    <label
      className="ms-Label ms-Toggle-label label-53"
      htmlFor="Toggle1"
    >
      IsChecked
    </label>
    <div
      className="ms-Toggle-innerContainer container-49"
    >
      <button
        aria-checked={false}
        checked={false}
        className="ms-Toggle-background pill-50"
        data-is-focusable={true}
        id="Toggle1"
        onChange={[Function]}
        onClick={[Function]}
        role="switch"
        type="button"
      >
        <div
          className="ms-Toggle-thumb thumb-51"
        />
      </button>
    </div>
  </div>
</form>
`;
